echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/TriggerRepository.groovy

```diff
@@ -0,0 +1,63 @@
+package com.netflix.spinnaker.echo.scheduler.actions.pipeline.impl
+
+import com.netflix.spinnaker.echo.model.Pipeline
+import com.netflix.spinnaker.echo.model.Trigger
+import groovy.util.logging.Slf4j
+
+import javax.annotation.Nullable
+
+@Slf4j
+class TriggerRepository {
+  private final Map<String, Trigger> triggersById;
+
+  TriggerRepository(List<Pipeline> pipelines) {
+    triggersById = new HashMap<>()
+    for (Pipeline pipeline : pipelines) {
+      for (Trigger trigger : pipeline.triggers) {
+        if (trigger.parent == null) {
+          trigger.parent = pipeline
```

Comment:
looking for feedback: I am not entirely happy with this. It feels gross to sort of hotpatch triggers after the fact like this, but I couldn't think of a better way to accomplish this. Note that once the dirty deed is done, it is quite nice to be able to look up `trigger.parent`

Reply:
I suppose you could also do something like the below inside of `PipelineCache` (aside from getting more complicated with json deserializers).

```
private void cachePipelines(List<Pipeline> pipelines) {
    log.info("Refreshing pipelines");

    pipelines = pipelines
      .stream()
      .map(p -> {
        List<Trigger> triggers = p.getTriggers()
          .stream()
          .map(t -> {
            t = t.withPipeline(p);
            t = t.withId("my_new_id");
            return t;
          })
          .collect(Collectors.toList());

        return p.withTriggers(triggers);
      })
      .collect(Collectors.toList());

    this.pipelines.set(pipelines);
  }
```

Reply:
My thinking would be along the lines of always generating a new id that is scoped within the `pipelineId` ... it could be a simple concatenation most likely, I'm not sure it _has_ to be a standard UUID.

Moving away from a persisted `id` in `front50`, I'm not sure there's much value in detecting where previously someone may have inadvertently re-used a trigger id across different pipelines?

Doing it within `PipelineCache` would keep things reasonable colocated and the rest of the logic would presumably not need to change much/at all?

Reply:
Doing it in `PipelineCache` seems sensible, it does provide a more consistent interface 👌